GVFS/GVFS.Common/Tracing/TelemetryDaemonEventListener.cs

```diff
@@ -0,0 +1,164 @@
+        {
+            // This listener is disabled unless the user specifies the proper git config setting.
+
+            string telemetryId = GetConfigValue(gitBinRoot, GVFSConstants.GitConfig.GVFSTelemetryId);
+            if (!string.IsNullOrEmpty(telemetryId))
+            {
+                return new TelemetryDaemonEventListener(providerName, enlistmentId, mountId, pipeName);
+            }
+            else
+            {
+                return null;
+            }
+        }
+
+        public override void Dispose()
+        {
+            this.pipeClient?.Dispose();
+            base.Dispose();
+        }
+
+        protected override void RecordMessageInternal(TraceEventMessage message)
+        {
+            string json = this.SerializeMessage(message);
+            this.SendMessage(json);
+        }
+
+        private static string GetConfigValue(string gitBinRoot, string configKey)
+        {
+            string value = string.Empty;
+            string error;
+
+            GitProcess.ConfigResult result = GitProcess.GetFromSystemConfig(gitBinRoot, configKey);
+            if (!result.TryParseAsString(out value, out error, defaultValue: string.Empty) || string.IsNullOrWhiteSpace(value))
+            {
+                result = GitProcess.GetFromGlobalConfig(gitBinRoot, configKey);
+                result.TryParseAsString(out value, out error, defaultValue: string.Empty);
+            }
+
+            return value.TrimEnd('\r', '\n');
+        }
+
+        private string SerializeMessage(TraceEventMessage message)
+        {
+            var pipeMessage = new TelemetryMessage
+            {
+                Version = this.vfsVersion,
+                ProviderName = this.providerName,
+                EventName = message.EventName,
+                EventLevel = message.Level,
+                EventOpcode = message.Opcode,
+                Payload = new TelemetryMessage.TelemetryMessagePayload
+                {
+                    EnlistmentId = this.enlistmentId,
+                    MountId = this.mountId,
+                    Json = message.Payload
+                },
+
+                // Other TraceEventMessage properties are not used
+            };
+
+            return pipeMessage.ToJson();
+        }
+
+        private void SendMessage(string message)
+        {
+            // Create pipe if this is the first message, or if the last connection broke for any reason
+            if (this.pipeClient == null)
+            {
+                var pipe = new NamedPipeClientStream(".", this.pipeName, PipeDirection.Out, PipeOptions.Asynchronous);
+
+                // Specify a instantaneous timeout because we don't want to hold up the rest of the
+                // application if the pipe is not available; we will just drop this event.
+                // We let any TimeoutExceptions bubble up and will try again with a new pipe on the next SendMessage call.
+                pipe.Connect(timeout: 0);

```

Comment:
Who is responsible for catching the `TimeoutException`? Does it risk bringing down the application if not caught?

Reply:
This is caught in the public method `EventListener.TryRecordMessage` which looks like this:

```csharp
public bool? TryRecordMessage(TraceEventMessage message, out string errorMessage)
{
    try
    {
        this.RecordMessageInternal(...);
    }
    catch (Exception ex)
    {
        errorMessage = ex.ToString();
        return false;
    }
}
```

The call stack is:
```
TelemetryDaemonEventListerner.SendMessage
TelemetryDaemonEventListerner.RecordMessageInternal
EventListener.TryRecordMessage
```

Reply:
Thanks!